hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/SqlNodeUtil.java

```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate;
+
+import com.hazelcast.sql.SqlErrorCode;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.type.converter.StringConverter;
+import org.apache.calcite.runtime.CalciteContextException;
+import org.apache.calcite.sql.SqlDynamicParam;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.math.BigDecimal;
+
+import static org.apache.calcite.sql.type.SqlTypeName.APPROX_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.CHAR_TYPES;
+import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
+import static org.apache.calcite.sql.type.SqlTypeName.EXACT_TYPES;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Utility methods to work with {@link SqlNode}s.
+ */
+public final class SqlNodeUtil {
+
+    private SqlNodeUtil() {
+    }
+
+    /**
+     * @return {@code true} if the given node is a {@linkplain SqlDynamicParam
+     * dynamic parameter}, {@code false} otherwise.
+     */
+    public static boolean isParameter(SqlNode node) {
+        return node.getKind() == SqlKind.DYNAMIC_PARAM;
+    }
+
+    /**
+     * @return {@code true} if the given node is a {@linkplain SqlLiteral literal},
+     * {@code false} otherwise.
+     */
+    public static boolean isLiteral(SqlNode node) {
+        return node.getKind() == SqlKind.LITERAL;
+    }
+
+    /**
+     * Obtains a numeric value of the given node if it's a numeric or string
+     * {@linkplain SqlLiteral literal}.
+     * <p>
+     * If the literal represents an exact value (see {@link
+     * SqlTypeName#EXACT_TYPES}), the obtained numeric value is {@link BigDecimal}.
+     * Otherwise, if the literal represents an approximate value (see {@link
+     * SqlTypeName#APPROX_TYPES}), the obtained numeric value is {@link Double}.
+     *
+     * @param node the node to obtain the numeric value of.
+     * @return the obtained numeric value or {@code null} if the given node is
+     * not a numeric or string literal.
+     * @throws CalciteContextException if the given node is a string literal
+     *                                 that doesn't have a valid numeric
+     *                                 representation.
+     */
+    public static Number numericValue(SqlNode node) {
+        if (node.getKind() != SqlKind.LITERAL) {
+            return null;
+        }
+
+        SqlLiteral literal = (SqlLiteral) node;
+        SqlTypeName typeName = literal.getTypeName();
+
+        if (CHAR_TYPES.contains(typeName)) {
+            try {
+                String value = literal.getValueAs(String.class);
+                if (value == null) {
+                    return null;
+                }
+
+                if (value.contains("e") || value.contains("E")) {
```

Comment:
What would happen if we have a literal that just happened to have `e` or `E` character, but otherwise is not a number? The exception mentions that the literal "cannot be converted to DECIMAL", while this is not really the case - we may have attempted to convert it to `DOUBLE`. Can we change it to something like `... cannot be converted to numeric type`?

Reply:
fixed in f19f51f2f0627047116118bc6af56d018ffac613